include/real/exact_number.hpp

```diff
@@ -693,190 +694,322 @@ namespace boost {
+                residual.normalize();
 
+                if(residual < zero){
+                    this->round_up(base);
+                }
 
-                // recalculate residual for the final (*this) value
-                exact_number<T> residual_o = ((*this) * divisor) - numerator;
+                if(residual > zero){
+                    this->round_down(base);
+                }
 
-                // note we have to normalize before comparison, because -0.0 != zero ..
-                residual_o.normalize();
+                this->positive = positive;
+                this->exponent += exponent_diff;
+                this->normalize();
+            }
 
-                if (residual_o != zero) { // then, we are not fully accurate
-                    // we try seeing if we can make the residual equal zero by adding/subtracting epsilon
-                    exact_number<T> tmp_lower = (*this);
-                    tmp_lower.round_down(base);
-                    exact_number<T> tmp_upper = (*this);
-                    tmp_upper.round_up(base);
+            /*          NEWTON RAPHSON DIVISION
+             * @brief:   calculates (*this)/divisor
+             * @param:   divisor: an exact_number which divides (*this)
+             * @param:   maximum_precision:Absolute Error in the result should be < 1*base^(-maximum_precision)
+             * @param:   upper: if upper is
+             *              1. true: returns result with [0, +epsilon] error
+             *              2. false: return result with [-epsilon, 0] error, here epsilon = 1*base^(-maximum_precision)
+             * @author:  Kishan Shukla
+             */
 
-                    residual = tmp_lower * divisor - numerator;
-                    residual.normalize();
+            void newton_raphson_division(
+                const exact_number<T> divisor,
+                unsigned int maximum_precision,
+                bool upper){
+
+                /* Exceptions start */
+                if (maximum_precision > (unsigned int) std::abs(std::numeric_limits<exponent_t>::min())) {
+                    throw exponent_overflow_exception();
+                }
 
-                    if (residual == zero) {
-                        (*this) = tmp_lower;
+                exact_number<T> zero = exact_number<T> ();
+                if(divisor == zero){
+                    throw divide_by_zero();
+                }
+                /* Exceptions end */
 
-                        if (positive)
-                            (*this).positive = true;
-                        else
-                            (*this).positive = false;
+                /* special cases start */
+                if((*this) == zero){
+                    return;
+                }
 
-                        (*this).normalize();
-                        this->exponent += exponent_dif;
-                        return;
-                    } 
+                exact_number<T> _1;
```

Comment:
please be consistent with the naming, above you used "one" and "minus_one" and "zero".

Reply:
I will change the names to follow a single system